cluster/kubernetes/namespacer.go

```diff
@@ -0,0 +1,166 @@
+package kubernetes
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	crdv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	crd "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/discovery"
+	discocache "k8s.io/client-go/discovery/cached"
+	toolscache "k8s.io/client-go/tools/cache"
+
+	kresource "github.com/weaveworks/flux/cluster/kubernetes/resource"
+)
+
+type namespaceViaDiscovery struct {
+	fallbackNamespace string
+	disco             discovery.DiscoveryInterface
+}
+
+type namespaceDefaulter interface {
+	GetDefaultNamespace() (string, error)
+}
+
+// NewNamespacer creates an implementation of Namespacer
+func NewNamespacer(ns namespaceDefaulter, d discovery.DiscoveryInterface) (*namespaceViaDiscovery, error) {
+	fallback, err := ns.GetDefaultNamespace()
+	if err != nil {
+		return nil, err
+	}
+	return &namespaceViaDiscovery{fallbackNamespace: fallback, disco: d}, nil
+}
+
+// effectiveNamespace yields the namespace that would be used for this
+// resource were it applied, taking into account the kind of the
+// resource, and local configuration.
+func (n *namespaceViaDiscovery) EffectiveNamespace(m kresource.KubeManifest) (string, error) {
+	namespaced, err := n.lookupNamespaced(m.GroupVersion(), m.GetKind())
+	switch {
+	case err != nil:
+		return "", err
+	case namespaced && m.GetNamespace() == "":
+		return n.fallbackNamespace, nil
+	case !namespaced:
+		return "", nil
+	}
+	return m.GetNamespace(), nil
+}
+
+func (n *namespaceViaDiscovery) lookupNamespaced(groupVersion, kind string) (bool, error) {
+	resourceList, err := n.disco.ServerResourcesForGroupVersion(groupVersion)
+	if err != nil {
+		return false, fmt.Errorf("error looking up API resources for %s.%s: %s", kind, groupVersion, err.Error())
+	}
+	for _, resource := range resourceList.APIResources {
+		if resource.Kind == kind {
+			return resource.Namespaced, nil
+		}
+	}
+	return false, fmt.Errorf("resource not found for API %s, kind %s", groupVersion, kind)
+}
+
+// This mainly exists so I can put the controller and store in here,
+// so as to have access to them for testing; and, so that we can do
+// our own invalidation.
+type cachedDiscovery struct {
+	discovery.CachedDiscoveryInterface
+
+	store      toolscache.Store
+	controller toolscache.Controller
+
+	invalidMu sync.Mutex
+	invalid   bool
+}
+
+// The older (v8.0.0) implementation of MemCacheDiscovery refreshes
```

Comment:
I am a bit confused. We are actually vendoring version v8.0.0 of client-go.

Reply:
Yes; if we were using the newer one (and there was no chance of going back, or it changing its implementation again), this wouldn't be necessary.

Reply:
Ah, older as compared to the latest version released not compared to the one vendored. Gotcha. I think some rephrasing will make it easier to understand though.